sdks/java/core/src/test/java/org/apache/beam/sdk/transforms/display/DisplayDataEvaluator.java

```diff
@@ -0,0 +1,146 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.beam.sdk.transforms.display;
+
+import org.apache.beam.sdk.Pipeline;
+import org.apache.beam.sdk.coders.Coder;
+import org.apache.beam.sdk.options.PipelineOptions;
+import org.apache.beam.sdk.options.PipelineOptionsFactory;
+import org.apache.beam.sdk.runners.PipelineRunner;
+import org.apache.beam.sdk.runners.TransformTreeNode;
+import org.apache.beam.sdk.transforms.Create;
+import org.apache.beam.sdk.transforms.PTransform;
+import org.apache.beam.sdk.values.PCollection;
+import org.apache.beam.sdk.values.POutput;
+
+import com.google.common.collect.Sets;
+
+import java.util.Objects;
+import java.util.Set;
+
+/**
+ * Test utilities to evaluate the {@link DisplayData} in the context of a {@link PipelineRunner}.
+ */
+public class DisplayDataEvaluator {
+  private final PipelineOptions options;
+
+  /**
+   * Create a new {@link DisplayDataEvaluator} using the specified {@link PipelineRunner} and
+   * default {@link PipelineOptions}.
+   */
+  public static DisplayDataEvaluator forRunner(Class<? extends PipelineRunner<?>> pipelineRunner) {
+    return forRunner(pipelineRunner, PipelineOptionsFactory.create());
+  }
+
+  /**
+   * Create a new {@link DisplayDataEvaluator} using the specified {@link PipelineRunner} and
+   * {@link PipelineOptions}.
+   */
+  public static DisplayDataEvaluator forRunner(
+      Class<? extends PipelineRunner<?>> pipelineRunner, PipelineOptions pipelineOptions) {
+    return new DisplayDataEvaluator(pipelineRunner, pipelineOptions);
+  }
+
+  private DisplayDataEvaluator(Class<? extends PipelineRunner<?>> runner, PipelineOptions options) {
+    this.options = options;
+    this.options.setRunner(runner);
+  }
+
+  /**
+   * Traverse the specified {@link PTransform}, collecting {@link DisplayData} registered on the
+   * inner primitive {@link PTransform PTransforms}.
+   *
+   * @return the set of {@link DisplayData} for primitive {@link PTransform PTransforms}.
+   */
+  public <InputT> Set<DisplayData> displayDataForPrimitiveTransforms(
+      final PTransform<? super PCollection<InputT>, ? extends POutput> root) {
+    return displayDataForPrimitiveTransforms(root, null);
+  }
+
+  /**
+   * Traverse the specified {@link PTransform}, collecting {@link DisplayData} registered on the
+   * inner primitive {@link PTransform PTransforms}.
+   *
+   * @param root The root {@link PTransform} to traverse
+   * @param inputCoder The coder to set for the {@link PTransform} input, or null to infer the
+   *                   default coder.
+   *
+   * @return the set of {@link DisplayData} for primitive {@link PTransform PTransforms}.
+   */
+  public <InputT> Set<DisplayData> displayDataForPrimitiveTransforms(
```

Comment:
We use "primitive transform" as a concept in a few places, but I didn't find instances of "leaf".

Does "leaf" accurately describe the concept? My understanding is that a primitive transform is an injection point where the runner may replace the node with a different implementation. Can the injected implementation include other transforms?